### ISSUE TITLE
[red-knot] port tests in infer.rs to mdtest

### ISSUE THREAD

**Issue by carljm** [astral-sh/ruff#13696]
All (or almost all?) tests currently in `crates/red_knot_python_semantic/src/types/infer.rs` should be ported to the new test framework (that is, to Markdown files in `crates/red_knot_python_semantic/resources/mdtest/` directory.)

Documentation of the new test framework is in `crates/red_knot_test/README.md` (or will be once https://github.com/astral-sh/ruff/pull/13695 lands.)

Generally we should try to group closely-related tests in the same Markdown file, but avoid having any one file become too large.

Some tests in `infer.rs` will require features we haven't added to the test framework yet. Those can be left alone for now (or we can add the feature to the test framework!)

Please comment here before starting work on any test porting, to avoid duplicated work.

**Comment by Lexxxzy** [comment]
Started to port, please review my initial steps

**Comment by Lexxxzy** [comment]
Also i have one question/proposal. What  to do with ignored tests? Maybe somehow I can add ignored to config, like:
````
```py path=b.py ignored="Reason"
class C: pass
```
````

**Comment by carljm** [comment]
For ignored tests, I would just include the test as a normal running test and assert whatever diagnostics/types we actually currently output for that case, but add a TODO in markdown prose text commenting on what the real behavior of the test should be.
